### ISSUE TITLE
Inconsistent results for overlap_slices between 1.0 and 1.1

### ISSUE THREAD

**Issue by saimn** [astropy/astropy#5086]
It seems that the behavior of `astropy.nddata.utils.overlap_slices` changed in 1.1 and it sometimes gives results that differ from Astropy 1.0 (when the small array shape is an odd number and the position's floating part is greater than 0.5):

1.0:

```
In [7]: overlap_slices((50,50), (21, 21), (30.6, 30.6))
Out[7]:
((slice(20, 41, None), slice(20, 41, None)),
 (slice(0, 21, None), slice(0, 21, None)))
```

1.1 and 1.2:

```
In [8]: overlap_slices((50,50), (21, 21), (30.6, 30.6))
Out[8]:
((slice(21, 42, None), slice(21, 42, None)),
 (slice(0, 21, None), slice(0, 21, None)))
```

Maybe related to #3727 ? 
ping @hamogu @larrybradley

**Comment by MSeifert04** [comment]
As far as I see it they changed it from casting to integer (discarding decimals) to (biased) rounding, that might explain this "off by one"-error. I haven't read all of the topic in #3333 but I got the impression it was subject to a lot of discussion.

**Comment by larrybradley** [comment]
I suspect this changed in #3333, but I'd need to go back a look at that very lengthy discussion.

**Comment by hamogu** [comment]
@saimn : As the author of #3333, I can confirm that this change is intentional and the result of the discussion in #3333.

**Comment by hamogu** [comment]
I should have probably be a little more explicit to save you the time to read the entire #3333 thread:
The old (astropy 1.0) behavior was plain wrong for negative numbers (it  treated them and indices, counting from the end of the array instead of extracting something that's centered slightly outside the array), and buggy for float (1,2,3, are considered pixel centers, so 2.999, 3, and 3.0001 should give you the same result - that's particularly important because floating point precision is limited and in some cases you might get 2.9999999999999999. This will be printed to screen as rounded "3.0" and should definitely give you the same result as "3").

This bug was fixed in #3333, but because I had to rebase #3333, I eventually ended up closing it and opening #3727 instead.

**Comment by saimn** [comment]
@hamogu - Ok, thanks for the clarification, I started reading #3333 but did not have the time to go through all of it. But then, if the behavior of 1.0 is wrong, would it be possible to backport a fix ? It is a LTS, and it is annoying to have different results when you need to support both 1.0 and 1.1/1.2.

**Comment by hamogu** [comment]
@astrofrog @embray : See comment above: Should #3727 be backported?

@saimn : Note that that commit did more than just fixing the error, it also introduced several modes (partial, fill, strict) which could be considered a new feature.

**Comment by astrofrog** [comment]
@hamogu - is there any way to extract purely the bug fix part of #3737, or do you think it's too tightly coupled to the new features? I think in this special case it might be worth doing a PR to the v1.0.x branch just to do a subset of the changes that were done in #3727 since I think we need to be strict in not introducing new features in the LTS branch.

**Comment by hamogu** [comment]
@astrofrog : It would be possible to extract that, but it's not a 5  min fix, it would take considerable work because I would have to go through all the index stuff again and that always makes my head swirl. I probably won't have time to do that in the next few months (and if we don't find anybody to fix that soon, then even the LTS phase for 1.0 will be over).

**Comment by mwcraig** [comment]
I think this can be closed since the 2.0 release is coming up....if you agree @saimn, can you please close it?

**Comment by saimn** [comment]
@mwcraig - Ok, closing.
